C++98 (stlport) support for core FlatBuffers and FlexBuffers.

stewartmiles opened the pull request:
* Added internal - limited - implementation of flatbuffers::unique_ptr
  for STLs that don't ship with std::unique_ptr.  In C++11 and beyond
  this is just an alias for std::unique_ptr.
* Aliased used type traits structs is_scalar is_floating_point is_unsigned
  into flatbuffers namespace so they can be replaced in C++98 implementations.
  Right now these point at stlport's TR1 implementations.
* Wrapped vector::data() in VectorData().
* Added variants of FlatBufferBuilder::CreateVector() and
  FlatBufferBuilder::CreateVectorOfStructs() that allow the use of plain
  function pointers.
  Generated code has also been modified to use plain functions to build objects
  rather than std::function() so all generated code will work in C++98
  applications.
* Added flexbuffers::Builder::Vector(), flexbuffers::Builder::TypedVector()
  and flexbuffers::Builder::Map() methods that allow the use of plain function
  pointers.
* Changed Parser to internally use plain function pointers when parsing table
  and vector delimiters.
* Wrapped vector::emplace_back() in VectorEmplaceBack().
* Added specializations of NumToString() for 64-bit types that aren't supported
  by stringstream in stlport.
* Overloaded numeric_limits for 64-bit types not supported by stlport.
* Replaced build_apk.sh (which was broken by deprecation of the
  "android" tool in the Android SDK) with build.gradle and the
  appropriate gradle wrapper to build an APK.
* Switched Android build to build against all STL variants.
* Updated travis configuration to build Android test and sample.

Tested:
* Verified all tests continue to work on Linux, OSX and Android.
* Verified Travis build is green.

Change-Id: I9e634363793f85b9f141d21454b10686020a2065

Thank you for submitting a PR!

Please make sure you include the names of the affected language(s) in your PR title.
This helps us get the correct maintainers to look at your issue.

If you make changes to any of the code generators, be sure to run
`cd tests && sh generate_code.sh` (or equivalent .bat) and include the generated
code changes in the PR. This allows us to better see the effect of the PR.

If your PR includes C++ code, please adhere to the Google C++ Style Guide,
and don't forget we try to support older compilers (e.g. VS2010, GCC 4.6.3),
so only some C++11 support is available.

Include other details as appropriate.

Thanks!

stewartmiles commented:
Looks like I need to get this working with VS2010 as well :/

alexames commented on include/flatbuffers/idl.h:
```diff
@@ -415,7 +419,7 @@ struct IDLOptions {
       protobuf_ascii_alike(false),
       lang(IDLOptions::kJava),
-      lang_to_generate(0) {}
+      lang_to_generate(0)   {}
```
nit:spaces

stewartmiles replied:
fixed

alexames commented on include/flatbuffers/util.h:
```diff
@@ -202,9 +217,11 @@ inline std::string ConCatPathFileName(const std::string &path,
-    if (filepath.back() == kPathSeparatorWindows) {
-      filepath.back() = kPathSeparator;
-    } else if (filepath.back() != kPathSeparator) {
+    // NOTE: This avoids std::string::back() as it's only available in C++11.
```
optional: Worth writing a StringBack() function?

stewartmiles replied:
sg, done

stewartmiles commented on include/flatbuffers/base.h:
```diff
@@ -173,5 +173,163 @@ inline size_t PaddingBytes(size_t buf_size, size_t scalar_size) {
+#else
+  return vector->data();
+#endif  // defined(FLATBUFFERS_CPP98_STL)
+}
```
sg, done

stewartmiles commented on include/flatbuffers/base.h:
```diff
@@ -173,5 +173,163 @@ inline size_t PaddingBytes(size_t buf_size, size_t scalar_size) {
+#else
+  vector->emplace_back(data);
+#endif  // defined(FLATBUFFERS_CPP98_STL)
+}
```
done

stewartmiles commented on include/flatbuffers/base.h:
```diff
@@ -173,5 +173,163 @@ inline size_t PaddingBytes(size_t buf_size, size_t scalar_size) {
+};
+#endif  // !(defined(_MSC_VER) && _MSC_VER <= 1700 /* MSVC2012 */)
+#else
+// Very limited implementation of unique_ptr.
```
Ok will do, I'll ping this comment when I have an update since this is a larger job.

stewartmiles commented on include/flatbuffers/base.h:
```diff
@@ -173,5 +173,163 @@ inline size_t PaddingBytes(size_t buf_size, size_t scalar_size) {
+template <class T>
+using unique_ptr = std::unique_ptr<T>;
+#else
+// MSVC 2010 doesn't support C++11 aliases.
```
This forwards unique_ptr into the flatbuffers namespace, I've updated the comment to describe the rationale.

stewartmiles commented on include/flatbuffers/flatbuffers.h:
```diff
@@ -1248,6 +1262,28 @@ class FlatBufferBuilder
+  template <typename T, typename F, typename S>
+  Offset<Vector<const T *>> CreateVectorOfStructs(size_t vector_size, F f,
+                                                  S *state) {
+    StartVector(vector_size * sizeof(T) / AlignOf<T>(), AlignOf<T>());
```
Apart from creating a pair of CreateVectorOfStructStart() / CreateVectorOfStructEnd() methods to handle the couple of lines of pre / post code around the loop.  I don't think so, that of course adds the potential for the compiler to not inline this code in addition makes it a little more tricky to read.  Any other suggestions?

aardappel replied:
I guess I'd be fine adding such functions, since if this call is ever speed-sensitive, that will be about the loop, not about the pre/post code. Also, a user will likely only ever use one of the two functions in a code base, so inlining is likely.

stewartmiles replied:
Ok done

stewartmiles commented on include/flatbuffers/idl.h:
```diff
@@ -406,7 +410,7 @@ struct IDLOptions {
       escape_proto_identifiers(false),
       generate_object_based_api(false),
-      cpp_object_api_pointer_type("std::unique_ptr"),
+      cpp_object_api_pointer_type("flatbuffers::unique_ptr"),
```
Can't do this as some users of stlport (we have some internal users) have defined std::unique_ptr as well so it's not safe to just inject the type into the std namespace.  flatbuffers::unique_ptr is an alias of std::unique_ptr in all implementations except for stlport and <= MSVC2012. 

In <= MSVC2012 the derived class is a *little* tricky as this results in the user doing...

flatbuffers::unique_ptr fbp;
std::unique_ptr sup(fbp.release());

which I concede is inconvenient.  AFAIK this is the closest to type aliasing that's possible with <= MSVC2012.

BTW It's most definitely not 99% of the world not caring about stlport I'm afraid.  I've pinged you the data.

I guess another way to look at this is why support MSVC2010 / MSVC2012?

stewartmiles replied:
I've been pondering this a bit more and perhaps I can handle stlport using macros in the generated code e.g

in stl_emulation.h do..

`
namespace std {
class flatbuffers_unique_ptr;
}
`
Then in the generated code...

`
// generated header
#ifdef FLATBUFFERS_CPP98_STL
#define unique_ptr flatbuffers_unique_ptr
#endif  // FLATBUFFERS_CPP98_STL

// generated code that uses unique_ptr
#ifdef FLATBUFFERS_CPP98_STL
#undef unique_ptr
#endif  // FLATBUFFERS_CPP98_STL
`

This is a little bit of a pain since all internal uses of std::unique_ptr will also need to be wrapped in a similar set of boilerplate which is pretty unfortunate.

aardappel replied:
I think having non-STLPort users be able to use `std` types is pretty important, so yes, lets go this way.

I'd love to drop VS2010, since the pressure of people wanting to use more of C++11 is pretty great. This is something we can consider seperately. But first order of business is not requiring them use even less of C++11 because of STLPort :)

stewartmiles commented on samples/android/gradlew:
```diff
@@ -0,0 +1,172 @@
+#!/usr/bin/env sh
```
Yep, that's the canonical way to distribute gradle based projects.

stewartmiles commented on src/idl_parser.cpp:
```diff
@@ -28,6 +28,79 @@
+  static unsigned long long max() { return ~0ULL; }
+};
+
+template <>
```
Sure, see earlier comment.  I'll ping when I've moved it.

stewartmiles replied:
Ok it's moved.

stewartmiles commented on src/idl_parser.cpp:
```diff
@@ -28,6 +28,79 @@
+// MSVC 2010 does not allow lambdas (with no captures) to be passed as plain
+// C++ functions so instatitate a wrapper that forwards to the lambda.
+template<typename F>
+class ParseDelimitersCallbackHolder {
```
Unfortunately it can't be a macro as a lambda in <= MSVC2012 can't just be cast to a function pointer it seems to be a pretty broken state of affairs.  I could split the logic out into static methods rather than using lambdas but it makes the code harder to follow WDYT?

aardappel replied:
Hard to tell which is worse without seeing the code, but begin/end style methods to factor out some of the common code maybe preferrable. I used a lambda because it was the simplest way to factor out the common code, but if a single use of a lambda requires all this helper code, it's not good value for money anymore.

stewartmiles replied:
Did you see the latest patch?  I've ditched this wrapper code in favor of #ifdef'in the typedef for the function pointer which is *far* cleaner.

stewartmiles commented on tests/test.cpp:
```diff
@@ -1566,24 +1571,24 @@ void FlexBuffersTest() {
-      slb += "Fred";
-      slb.IndirectFloat(4.0f);
-    });
+  slb.Map([](flexbuffers::Builder& slb2) {
```
I still wanted to make sure the new code worked.  I guess I could add a version that works with old skool function pointers and new lambda / closures, WDYT?

aardappel replied:
yes, a duplicate test under an #ifdef would be better.

stewartmiles commented:
Responses to the top level comment inline below:

Thanks for the mountain of work to support STLPort. I've got some nits and one bigger issue:

So far we've promised that STLPort is supported for flatbuffers.h and the generated code (without --object-api), since otherwise the changes would be too sprawling (as you just found out). We never enforced this, which your change now does, which is awesome.

Is your project using STLPort also using the parser, flexbuffers, reflection, and the object API?

>> We're using the parser, but the cost of adding support for the other components was minimal.

If not, I feel this change could have been simplified a lot by simply making a new test target for Android STLPort that only includes the basics.

>> I disagree, supporting STLport requires - TBH - minimal changes to the code base.  Most of the C++11 features utilized are simply syntactic sugar that only serve to narrow the user base of this library and components that depend upon it.

While it could be argued that STLPort compatibility is a "nice to have" for the entire code-base, it puts undue burden on contributors to have to support it. They typically want to use full C++11 in their changes, and then have to deal with CI telling them it is not supported on VS2010. Luckily, VS2010 supports a lot of the basics, but STLPort is much worse, also because most contributors won't have a local test-setup for it, so are forced to play a very slow ping-pong with the CI to get their PR in shape.

>> The tests are already in place.  Android CI is now setup for this, VS2010 is humming along the limits on STL usage are minimal.

And all that for a build config that is dying out rapidly, and very few people are using. I feel it should only be supported where absolutely necessary, and impact the modern C++ experience as little as possible.

>> See the data I forwarded you, basically it disagrees strongly with your statement.

stewartmiles commented on include/flatbuffers/util.h:
```diff
@@ -60,6 +60,21 @@ template<> inline std::string NumToString<signed char>(signed char t) {
+#if defined(FLATBUFFERS_CPP98_STL)
+template <>
+inline std::string NumToString<long long>(long long t) {
+  char buf[21]; // (log((1 << 63) - 1) / log(10)) + 2
```
How would we do that?  log AFAIK is not constexpr.  Even if it had constexpr support in some STLs what about those that don't?

aardappel replied:
Sorry, I was under the assumption the compiler would tolerate `log` in regular const expressions, that indeed doesn't make sense.

stewartmiles commented:
Ok:
* I've moved the cruft into stl_emulation.h 
* Renamed VectorData etc. methods.
* Cleaned up flatbuffers::unique_ptr in <= MSVC2012
* Restored the std::function code that constructs flexbuffers in the test.  Though I did narrow the capture to just the builder, it's *interesting* how similar the code looks.

aardappel commented:
Then I would suggest we make a new android_stlport target, that only uses the basic generated code (without `--object-api`) and the parser, and revert all STLPort changes to the rest of the code base.

Like I said, it is not about the cost in code for you to make these changes, it is about the cost to future PRs having to deal with writing STLPort compatible code and using our wrappers correctly, relying on CI to tell them. I believe that burden should be as minimal as possible.

In the case of the object API in particular, we make the promise that this translates FlatBuffers into objects using "idiomatic C++", so much so that you can use them as your runtime data structures. If suddenly something as basic as `std::unique_ptr` becomes a different type, that can polute into downstream code or require conversions, which defies the point of this being idiomatic. So if the object API has to be made STLPort compatible, I'd suggest #ifdefs in the generated code such that non-STLPort users can still have their `std` types. But better yet let's not have this compatibility for the object API at all, since it is optional anyway.

stewartmiles commented:
I agree that std::unique_ptr --> flatbuffers::unique_ptr with <= MSVC2012 is suboptimal.  I strongly disagree that about your assessment associated with maintaining stlport compatibility, most of this CL is build logic and stl_emulation.h.

w.r.t unique_ptr in the object API code gen, how about I move flatc's default back to using std::unique_ptr and I change the build logic for the tests / samples to generate code with flatbuffers::unique_ptr?  This will keep the non-stlport code gen using std::unique_ptr so that we can support <= MSVC2012 in a canonical way (note this is the only affected compiler by flatbuffers::unique_ptr vs. std::unique_ptr).  WDYT?

stewartmiles commented:
Ok by default the object API generates code with std::unique_ptr now.

aardappel commented:
Seems the only main disagreement we have left is wether STLPort compatibility should be inflicted on parts on the code base that currently don't need it. Can we have some other people weigh in?

aardappel commented on include/flatbuffers/base.h:
```diff
@@ -29,6 +29,8 @@
   #include <functional>
 #endif
 
+#include "flatbuffers/stl_emulation.h"
```
My intent was that this `#include` be conditional on `#ifdef FLATBUFFERS_CPP98_STL`, such that the file is generally not needed at all.

stewartmiles replied:
Not possible, as this uses type aliases for internal uses of unique_ptr etc.

aardappel commented on include/flatbuffers/flatbuffers.h:
```diff
@@ -1470,6 +1502,21 @@ class FlatBufferBuilder
+  // Allocates space for a vector of structures.
+  // Must be completed with EndVectorOfStructs().
+  template<typename T>
+  const T* StartVectorOfStructs(size_t vector_size) {
```
Should these be private?

stewartmiles replied:
Sure done, though I noticed a load of other stuff under protected that probably shouldn't be there.  Probably worth a future clean up.

aardappel commented on include/flatbuffers/flatbuffers.h:
```diff
@@ -1470,6 +1502,21 @@ class FlatBufferBuilder
 
+  // Allocates space for a vector of structures.
+  // Must be completed with EndVectorOfStructs().
+  template<typename T>
```
I believe the majority of occurrences of `template` in this code base put them on the same line.

stewartmiles replied:
Correct, the existing use cases are in violation of Google style.  Rather than proliferate the problem I've added code that adheres to Google style.  The idea is that a follow up can consist of a pass to align style so that we can use auto formatting tools like clang-format to keep things in check.

stewartmiles replied:
Sure, done.  Though I notice that we have a load of other stuff under protected that probably shouldn't be there.

aardappel replied:
See my comment elsewhere.. consistency > any style guide. If this is to be fixed, then all at once, and not in this PR.

stewartmiles replied:
Ok we'll do that.

aardappel commented on include/flatbuffers/flexbuffers.h:
```diff
@@ -743,7 +745,7 @@ inline Reference GetRoot(const uint8_t *buffer, size_t size) {
 
 inline Reference GetRoot(const std::vector<uint8_t> &buffer) {
-  return GetRoot(buffer.data(), buffer.size());
+  return GetRoot(flatbuffers::vector_data(&buffer), buffer.size());
```
In my last review I asked if this could be called `data` (to match the STL function). It already sits in the `flatbuffer` namespace, so I don't feel this needs `vector_` to be clear.
Also, I think the argument should be by `&` rather than `*`, to match STL, and since most uses are const anyway.

stewartmiles replied:
Regarding

vector_data(vector* v) vs. vector_data(const vector& v)

there are mutable uses of vector.data()  Since it's possible for the result to be mutable, following Google style, it can't be a reference.  This makes it clear at the point of use that the data could be mutated.  The alternative would be to have

T* vector_data(vector<T>* v) 
T* vector_data(const vector& v)

which I feel is worse.  If you want me to go that way, happy to do so.

aardappel replied:
We're replacing the STL, so STL style > Google style in this case. So both `&` would be best. Failing that, just the const one.

stewartmiles replied:
I've replaced everything with a reference though I *really* don't like it.  It simply aids to obfuscate the code at the call site.  This is something that could be resolved during a style pass.

aardappel commented on include/flatbuffers/idl.h:
```diff
@@ -571,15 +575,30 @@ class Parser : public ParserState {
   FLATBUFFERS_CHECKED_ERROR ParseAnyValue(Value &val, FieldDef *field,
                                           size_t parent_fieldn,
                                           const StructDef *parent_struct_def);
+#if defined(FLATBUFFERS_CPP98_STL)
```
indent similarly to the rest of the project

stewartmiles replied:
Done, we'll resolve in a follow up.

aardappel commented on include/flatbuffers/stl_emulation.h:
```diff
@@ -0,0 +1,228 @@
+
+template <typename T, typename V>
+inline void vector_emplace_back(std::vector<T> *vector, V &&data) {
+#if defined(FLATBUFFERS_CPP98_STL)
```
indent similarly to rest of project, here and elsewhere

stewartmiles replied:
I'm afraid that would be a Google style violation.  I can see the case for indentation where there is a lot of nested preprocessor logic, however this is a very clear case.

aardappel replied:
We're replacing the STL, so STL style > Google style in this case. So both `&` would be best. Failing that, just the const one.

aardappel replied:
Then the whole project is in violation. Consistency is more important than adhering to a style however, so this should be indented until such time that the whole project is fixed, if that is to happen. I don't think it is a good idea to leave the project in an inconsistent state.

stewartmiles replied:
Ok, we can resolve in a follow up.

alexames commented:
Does this sound like a workable idea? What if instead of adding functions like flatbuffers::vector_back(), we added a flatbuffers::vector class - under everything except STLPort this is just:

    template <typename T> using vector = std::vector<T>;

but under STLPort it's a class that extends std::vector and implements the functions we care about (data, back, emplace_back, all the things stewart already wrote wrappers for). Then the only things that need to change are the types of the data. The code itself can be written in idiomatic C++11. (It's not a perfect solution of course, if we decide to use other C++11 features we have to add those to our customized version of the classes, but the number of new functions on the container classes is relatively small I think.)

As for the things new to C++11 like unique_ptr, I feel like we should be able cheat a little bit and just drop those in namespace std even though that's technically UB. (And only under STLPort builds of course). 

If the above ideas are not feasible solutions, I'm not sure whether it's worse to increase the hurdles to contributing or lose out on potential users who are married to STLPort. I know there's a surprising number of people using STLPort but I don't know how many of them are itching to use flatbuffers.

aardappel commented:
@alexames yup, that would actually be better, since it reduces the friction with people not familiar with STLPort. Maybe nice for a followup-PR?

One risk is that if people use a new function not available in this emulation class (but generally available even in STLPort), they'll now be forced to add to the emulation class, whereas before that just worked. So the emulation better be somewhat complete.

aardappel commented:
Ok, I guess this can be merged if everyone is happy.

stewartmiles commented:
@alexames and @aardappel , extending std::vector and std::string in stl_emulation.h sounds like a neat idea.  This does have the issue that flatbuffers::vector != std::vector when using stlport which means that code like `std::vector v = someflatbuffersvector` will fail to compile as there is no conversion between the derived class and the base class.  This is a similar issue to the horribleness that occurs when using flatbuffers::unique_ptr in <= MSVC2012 builds.  On the up side this does limit the pain to users of stlport - who should move away from it anyway.  If you folks think I should make the change, I'll do so.

@alexames moving flatbuffers::unique_ptr into the std:: namespace isn't feasible as I've already - unfortunately - run across some projects using stlport that decide to do exactly the same resulting in duplicate type definitions for std::unique_ptr.

stewartmiles commented:
I took a first stab at moving to flatbuffers::vector and flatbuffers::string to get rid of the compatibility methods and the change ends up inconsistently bubbling up into the public API so we end up with a load of occasional flatbuffers::vector and flatbuffers::string in the API surface.  In order to do this properly we would need to change all public APIs to use flatbuffers::vector or flatbuffers::string instead which is a far larger change.  I would definitely prefer to keep what we have in place and add custom vector / string types in a follow up commit.
